// nuxt-anchorscroll@1.0.5/dist/module.mjs
import { defineNuxtModule, createResolver, addPlugin, addImportsDir } from '@nuxt/kit';

const name = "nuxt-anchorscroll";
const type = "module";
const version = "1.0.5";
const packageManager = "yarn@3.6.1";
const description = "Scroll with anchor support for Nuxt";
const license = "MIT";
const repository = "helltraitor/nuxt-anchorscroll";
const keywords = [
	"anchor",
	"nuxt",
	"nuxt-module",
	"scroll",
	"vue"
];
const exports = {
	".": {
		types: "./dist/types.d.ts",
		require: "./dist/module.cjs",
		"import": "./dist/module.mjs"
	}
};
const main = "./dist/module.cjs";
const types = "./dist/types.d.ts";
const files = [
	"dist"
];
const scripts = {
	prepack: "nuxt-module-build build",
	dev: "nuxi dev playground",
	"dev:build": "nuxi build playground",
	"dev:prepare": "nuxt-module-build --stub && nuxi prepare playground",
	release: "npm run lint && npm run prepack && npm publish && git push --follow-tags",
	lint: "eslint ."
};
const dependencies = {
	"@nuxt/kit": "^3.5.2"
};
const devDependencies = {
	"@antfu/eslint-config": "^0.39.7",
	"@iconify-json/carbon": "^1.1.18",
	"@iconify/json": "^2.2.88",
	"@nuxt/content": "^2.7.0",
	"@nuxt/devtools": "latest",
	"@nuxt/eslint-config": "^0.1.1",
	"@nuxt/module-builder": "^0.8.3",
	"@nuxt/schema": "^3.15.0",
	"@nuxt/test-utils": "^3.5.1",
	"@types/node": "^18",
	"@unocss/nuxt": "^0.53.4",
	changelogen: "^0.5.3",
	eslint: "^8.42.0",
	nuxt: "^3.12.4",
	sass: "^1.63.6",
	vitest: "^0.31.4",
	"vue-router": "^4.2.4"
};
const pkg = {
	name: name,
	type: type,
	version: version,
	packageManager: packageManager,
	description: description,
	license: license,
	repository: repository,
	keywords: keywords,
	exports: exports,
	main: main,
	types: types,
	files: files,
	scripts: scripts,
	dependencies: dependencies,
	devDependencies: devDependencies
};

const module = defineNuxtModule({
  meta: {
    name: "nuxt-anchorscroll",
    configKey: "anchorscroll",
    compatibility: {
      nuxt: ">=3.0.0"
    },
    version: pkg.version
  },
  defaults: {
    hooks: ["page:finish"]
  },
  setup(options, nuxt) {
    const resolver = createResolver(import.meta.url);
    nuxt.options.runtimeConfig.public.anchorScroll = { ...options };
    addPlugin(resolver.resolve("./runtime/anchorscroll-plugin"));
    addImportsDir(resolver.resolve("./runtime/composables"));
  }
});

export { module as default };
